File: Execution/affordance_corrections/nodes/affordance_corrections/affordance_helpers/cppfitting/registration.cpp
// File name:  cppregistration.cpp
// Description: main registration routine called by python/cython to perform parallel ICP registration of models
// Authors: Mike Hagenow, Kevin Macauley, Nicole Gathman
// Date: 11/05/21

// File name:  cppregistration.cpp
// Description: main registration routine called by python/cython to perform parallel ICP registration of models
// Authors: Mike Hagenow, Kevin Macauley, Nicole Gathman
// Date: 11/05/21

#include <stdio.h>
#include <stdlib.h>
#include <string>
#include "Model.h"
//#include "urdf_helpers.cpp"
#include <iostream>
#include <Eigen/Core>
#include <Eigen/SVD>
#include <Eigen/Geometry> 
#include <vector>
#include <iterator>
#include <chrono>
#include <thread>
#include <pico_tree/eigen.hpp>
#include <pico_tree/kd_tree.hpp>
#include <omp.h>
#include <nlopt.hpp>

using namespace std;

// int counter {};

struct closest_alignment_model{
    Eigen::Matrix3Xd s_vals;
    Eigen::Matrix3Xd t_vals;
    Eigen::VectorXd dists;
    std::vector<int> close_enough_inds;

    closest_alignment_model(Eigen::Matrix3Xd s, Eigen::Matrix3Xd t, std::vector<double> d, std::vector<int> c){
        s_vals=s;
        t_vals=t;
        dists=Eigen::Map<Eigen::VectorXd, Eigen::Unaligned>(d.data(), d.size());
        close_enough_inds=c;
    }
};

struct opt_data_registration_residual{
    std::vector<double> x;
    Model* model;
    Eigen::Matrix3Xd t_vals;
    std::vector<int> close_enough_inds;
    Eigen::MatrixXd weights;
    Eigen::Matrix3Xd R_final;
    Eigen::Matrix3Xd t_final;
    opt_data_registration_residual(std::vector<double> x_temp,
                                    Model* model_temp,
                                    Eigen::Matrix3Xd t_vals_temp,
                                    std::vector<int> close_enough_inds_temp,
                                    Eigen::MatrixXd weights_temp,
                                    Eigen::Matrix3Xd R_final_temp,
                                    Eigen::Matrix3Xd t_final_temp){
        x=x_temp;
        model=model_temp;
        t_vals=t_vals_temp;
        close_enough_inds=close_enough_inds_temp;
        weights=weights_temp;
        R_final=R_final_temp;
        t_final=t_final_temp;
    }
};


struct icp_model{
    Eigen::Matrix3Xd R;
    Eigen::Matrix3Xd t;
    std::vector<double> angles;
    double scale;
    double error;

    icp_model(Eigen::Matrix3Xd R_final, Eigen::Matrix3Xd t_final, std::vector<double> angles_final, double scale_final, double error_final){
        R=R_final;
        t=t_final;
        angles=angles_final;
        scale=scale_final;
        error=error_final;        
    }
};

// Function for aligning closest points between source and target to be used for SVD
closest_alignment_model get_closest_alignment(Eigen::Matrix3Xd aligned_source, Eigen::Matrix3Xd target, pico_tree::KdTree<pico_tree::EigenTraits<Eigen::Matrix3Xd>>* target_tree, double max_distance_threshold){
    pico_tree::Neighbor<int, Eigen::Matrix3Xd::Scalar> nn;
    std::vector<int> indices;
    std::vector<int> indices_orig;
    std::vector<double> dists; 
    for(int i=0; i< aligned_source.cols(); i++){
        target_tree->SearchNn(aligned_source.col(i), &nn);  
        if(nn.distance<max_distance_threshold){
            indices.push_back(nn.index);
            indices_orig.push_back(i);
            dists.push_back(nn.distance);
        }
    }
    return closest_alignment_model(aligned_source(Eigen::placeholders::all,indices_orig), target(Eigen::placeholders::all, indices), dists, indices_orig);
}

// returns the rotation matrix from the x vector
Eigen::Matrix3Xd get_R_from_X(const std::vector<double> &x){
    Eigen::Quaterniond q;
    q.x() = x[0];
    q.y() = x[1];
    q.z() = x[2];
    q.w() = x[3];
    return q.normalized().toRotationMatrix();
}

// returns the translation matrix from the x vector
Eigen::Matrix3Xd get_trans_from_X(const std::vector<double> &x){
    Eigen::Matrix3Xd t = Eigen::MatrixXd::Zero(3,1);

    for(int i = 4; i < 7;i++){
        t(i-4,0) = x[i];
    }
    return t;
}

// returns the angles vector from the x vector
std::vector<double> get_angles_from_x(const std::vector<double> &x){
    std::vector<double> angles_final;
    for(size_t i = 7; i < x.size();i++){ // TODO: missing the scale
        angles_final.push_back(x[i]);
    }
    return angles_final;
}

// Calculates just the residual for a set of inputs (used both for residual and gradient calculations)
double calc_residual(const std::vector<double> &x, void *data){
    // counter++;
    opt_data_registration_residual *d = (opt_data_registration_residual *) data;

    Eigen::Matrix3Xd R = get_R_from_X(x);

    if(isnan(R(0,0))){
        printf("X: %f %f %f %f %f\n",x[0],x[1],x[2],x[3],x[4]);
    }

    Eigen::Matrix3Xd t = get_trans_from_X(x);
    std::vector<double> angles_final = get_angles_from_x(x);

    // If scaling allowed, it is the last element of the optimization
    double scale;
    d->model->scaling_allowed ? scale = x[-1] : scale = 1.0;

    Eigen::Matrix3Xd R_temp = R*d->R_final;
    Eigen::Matrix3Xd trans_temp = R*d->t_final + t;

    Eigen::Matrix3Xd aligned_source = d->model->getPointsWithAngles(angles_final, R_temp, trans_temp, scale);

    Eigen::Matrix3Xd s_vals = aligned_source(Eigen::placeholders::all,d->close_enough_inds);
    Eigen::VectorXd non_weighted_residual = (s_vals-d->t_vals).colwise().norm();
    Eigen::VectorXd weights_vec = d->weights.col(0);
    double residual = (weights_vec.dot(non_weighted_residual))/scale;
    // std::cout << "Iteration: " << counter << " residual: " << residual << std::endl;
    return residual;
}

//  Calculates a residual and gradient for nonlinear fitting of R, t, angles, and scale if applicable 
double registration_residual(const std::vector<double> &x, std::vector<double> &grad, void *data){
    

    // // Numerically approximate the gradient
    // double dirac = 1e-4;
    // if (!grad.empty()) {
    //     for(int i=0;i<int(x.size());i++){
    //         std::vector<double> temp = x;
    //         std::vector<double> temp2 = x;
    //         temp[i] -= dirac;
    //         temp2[i] += dirac;
    //         double bwd = calc_residual(temp,data);
    //         double fwd = calc_residual(temp2,data);
    //         grad[i] = (fwd-bwd)/(2*dirac);
    //     }
    // }
    
    // Calculate residual for current values
    return calc_residual(x,data);
}

//  quaternion must have magnitude 1 -- used in NL optimization 
double quat_con(const std::vector<double> &x, std::vector<double> &grad, void *data){
    double dirac = 1e-4;
    
    if (!grad.empty()) {
        for(int i=0;i<4;i++){
            std::vector<double> temp = x;
            temp[i]-=dirac;
            Eigen::Quaterniond q;
            q.x() = temp[0];
            q.y() = temp[1];
            q.z() = temp[2];
            q.w() = temp[3];
            double bwd = 1.0 - q.norm();
            temp[i]+=2.0*dirac;
            q.x() = temp[0];
            q.y() = temp[1];
            q.z() = temp[2];
            q.w() = temp[3];
            double fwd = 1.0 - q.norm();
            grad[i] = (fwd-bwd)/(2*dirac);
        }
    }
    // TODO: get rid of eigen
    Eigen::Quaterniond q;
    q.x() = x[0];
    q.y() = x[1];
    q.z() = x[2];
    q.w() = x[3];
    return 1.0 - q.norm();
}

// Take two arrays of points in 3d space and computes the registration.
//  source refers to mesh/model to be fit, target refers to scene/point cloud
//  source and target dimensions are 3xa and 3xb
//  Note: target should be a region of interest (~ centered wrt the selected point) 
icp_model icp(Model model, 
            Eigen::Matrix3Xd target, 
            pico_tree::KdTree<pico_tree::EigenTraits<Eigen::Matrix3Xd>>* target_tree, 
            int n,
            bool artic_svd_inital = false,
            Eigen::Matrix3Xd R_initial = Eigen::Matrix3d::Identity(),
            Eigen::Matrix3Xd t_initial = Eigen::MatrixXd::Zero(3,1),
            std::vector<double> angles_initial = {},
            double scale_initial = 1.0,
            double diameter = 1.0){

    bool converged = false;
    double max_distance_threshold = diameter;
    int num_iters = 0;
    
    // For the idependent values calculated in one run of the loop 
    Eigen::MatrixXd R;
    Eigen::MatrixXd trans;

    // Updated after every run of the loop
    Eigen::Matrix3Xd R_final = R_initial;
    Eigen::Matrix3Xd trans_final = t_initial;
    std::vector<double> angles_final = angles_initial;
    double prev_error = std::numeric_limits<double>::max();
    double scale_final = scale_initial;

    // Get initial sampling with angles
    Eigen::Matrix3Xd aligned_source = model.getPointsWithAngles(angles_final, R_final, trans_final, scale_final);

    // calculate the initial alignment
    closest_alignment_model ca = get_closest_alignment(aligned_source, target, target_tree, max_distance_threshold);
    
    if(ca.s_vals.cols()<3){
        // insufficient points for registration, return initial pose
        return icp_model(R_final,trans_final,angles_final,scale_final,std::numeric_limits<double>::max());
    }

    while(!converged && num_iters<n){
        // Weighting is based on the inverse of distances (with some normalization)
        Eigen::MatrixXd weights = (1.0+(ca.dists.array())).pow(-1).matrix();
        Eigen::MatrixXd weights_matrix = weights.asDiagonal();
        
        if(!model.requires_NL_fitting() or artic_svd_inital){ // use SVD if not fitting articulations/scale
            // Following this formulation: https://igl.ethz.ch/projects/ARAP/svd_rot.pdf
            Eigen::MatrixXd w_cent_s = (ca.s_vals*weights)/weights.sum();
            Eigen::MatrixXd w_cent_t = (ca.t_vals*weights)/weights.sum();
            
            for(int i = 0; i < ca.s_vals.cols(); i++){
                ca.s_vals.col(i)-=w_cent_s;
            }
            for(int i = 0; i < ca.t_vals.cols(); i++){
                ca.t_vals.col(i)-=w_cent_t;
            }

            Eigen::MatrixXd S = ca.s_vals * weights_matrix * ca.t_vals.transpose();
            Eigen::JacobiSVD<Eigen::Matrix3Xd> svd(S, Eigen::ComputeFullU | Eigen::ComputeFullV);
            Eigen::Matrix3Xd V = svd.matrixV();

            if(svd.matrixU().determinant()*V.determinant()<0){
                for(int i = 0; i<V.rows();i++){
                    V(i,2) *= -1;
                }
            }

            R = V*svd.matrixU().transpose();
            trans = (w_cent_t - (R * w_cent_s));

            R_final = R * R_final;
            trans_final = R * trans_final + trans;

        } else {    // articulations, need to use NL optimization
            // structure of minimization: R as quaternion, t as (3x), angles as list
            // initialize parameters


            std::vector<double> x0((7+angles_final.size()+int(model.scaling_allowed)), 0); // 2nd arg -- initialize to zero
            std::vector<double> quat_temp{0.0, 0.0, 0.0, 1.0};

            // Add initial quaternion
            for(int i = 0; i < 4; i++){
                x0[i] = quat_temp[i];
            }

            // Add initial translation
            for(int i = 4; i < 7;i++){
                x0[i] = 0.0;
            }

            // Add initial angles
            for(size_t i = 7; i < 7+angles_final.size();i++){
                x0[i] = angles_final[i-7];
            }

            // bounds for pose (quaterions and translation)
            // Lower bound
            std::vector<double> lb{-1.0, -1.0, -1.0, -1.0, -model.diameter/2, -model.diameter/2, -model.diameter/2};
            // Upper bound
            std::vector<double> ub{1.0, 1.0, 1.0, 1.0, model.diameter/2, model.diameter/2, model.diameter/2};
            
            // Add bounds for all articulations
            for(size_t i = 0; i < angles_final.size(); i++){
                lb.push_back(model.joints[i].min_angle);
                ub.push_back(model.joints[i].max_angle);
            }

            // Add bounds for scaling if enabled
            if(model.scaling_allowed){
                lb.push_back(model.min_scale);
                ub.push_back(model.max_scale);
            }
    
            opt_data_registration_residual data_min_obj(x0, &model, ca.t_vals, ca.close_enough_inds, weights, R_final, trans_final);
            
            nlopt::opt opt(nlopt::algorithm::LN_COBYLA, x0.size());
            opt.set_lower_bounds(lb);
            opt.set_upper_bounds(ub);
            opt.set_min_objective(registration_residual, &data_min_obj);
            opt.set_maxeval(500);
            opt.add_equality_constraint(quat_con, NULL, 0.0001);
            opt.set_ftol_rel(0.0001);
            // counter = 0;
            double minf;

            try{
                nlopt::result result = opt.optimize(x0, minf);
                // cout << "RESULT: " << result << endl;
                if(result < 1){ // positive values are successful
                    return icp_model(R_final,trans_final,angles_final,scale_final,std::numeric_limits<double>::max());
                }
            }
            catch(nlopt::roundoff_limited const&){
                cout << "roundoff limited" << endl;
            }

            // Capture output from minimization
            R = get_R_from_X(x0);
            trans = get_trans_from_X(x0);
            angles_final = get_angles_from_x(x0);

            if(model.scaling_allowed){
                scale_final = x0[x0.size()-1];
            }

            R_final = R * R_final;
            trans_final = R * trans_final + trans;
        }

        // Transform source points
        aligned_source = model.getPointsWithAngles(angles_final,R_final, trans_final, scale_final);
        
        // Check if done - calculate residual based on next set of pt alignments
        ca = get_closest_alignment(aligned_source,target,target_tree,max_distance_threshold);
        // need 3 points
        if(ca.s_vals.cols()<3){ 
            // insufficient points for registration, return initial pose
            // return as failure (infinite error)
            std::cout << "Exiting Early" << std::endl;
            return icp_model(R_final,trans_final,angles_final,scale_final,std::numeric_limits<double>::max());
        }

        // // Calculate residual
        Eigen::Matrix3Xd t_vals_adjust = ca.t_vals;

        for(int i = 0; i < ca.t_vals.cols(); i++){
            t_vals_adjust.col(i) = trans - t_vals_adjust.col(i);
        }
        double error  = ((R * ca.s_vals) + t_vals_adjust).norm();
        
        // Check if converged (changing by small amount -- empiricially determined)
        double error_per_pt = 0.000000025;
        if(std::abs(prev_error-error)<(error_per_pt*ca.s_vals.cols())){
            converged = true;
        }
        prev_error = error;
        num_iters++;
    }
    return icp_model(R_final,trans_final,angles_final,scale_final, prev_error);
}

// Run ICP multiple times to avoid local minima (random rotations, translations, angles, scale).
icp_model icp_random_restarts(Model model, 
                                Eigen::Matrix3Xd target,
                                int num_restarts,
                                int n,
                                double diameter,
                                Eigen::Vector3d ref_point,
                                bool artic_svd_inital){
    
    double min_error = std::numeric_limits<double>::max();

    // initial value for scale
    double scale_final = 1.0;

    // Store target in KDTree for fast lookup of closest points
    int const kMaxLeafCount = 16;
    pico_tree::KdTree<pico_tree::EigenTraits<Eigen::Matrix3Xd>> tree(target, kMaxLeafCount);
    pico_tree::KdTree<pico_tree::EigenTraits<Eigen::Matrix3Xd>> *treePtr = &tree;

    // Default return values
    Eigen::Matrix3Xd R_final = Eigen::Matrix3d::Identity();
    Eigen::Matrix3Xd t_final(3,1);
    t_final.col(0) = ref_point;
    std::vector<double> angles_final(model.joints.size(), 0.0);

    icp_model icp_model_final = icp_model(R_final,t_final,angles_final,scale_final, min_error);

    std::default_random_engine generator;
    std::normal_distribution<double> distribution(0.0,0.33);

    for(int ii = 0; ii < num_restarts; ii++){
        // std::cout<<"RR: " << ii <<std::endl;
        // Get random start for orientation, translation, and all relevant ariculations
        Eigen::Quaterniond q = Eigen::Quaterniond::UnitRandom();
        Eigen::Matrix3Xd R_init = q.normalized().toRotationMatrix();
        q = Eigen::Quaterniond::UnitRandom();
        Eigen::Matrix3Xd t_init(3,1);
        t_init.col(0) = ((diameter/2.0)*distribution(generator)*q.normalized().toRotationMatrix().row(0).transpose()+ref_point);

        // Model articulation
        std::vector<double> angles;
        for(size_t jj = 0; jj < model.joints.size(); jj++){
            angles.push_back(model.joints[jj].min_angle + (model.joints[jj].max_angle-model.joints[jj].min_angle)*((double)rand()/(double)RAND_MAX));
        }

        double scale_init = 1.0;
        if(model.scaling_allowed){
            scale_init = model.min_scale + (model.max_scale-model.min_scale)*((double)rand()/(double)RAND_MAX);
        }

        icp_model icp_model_temp = icp(model,target,treePtr,n, artic_svd_inital, R_init, t_init, angles, scale_init, diameter);

        if(icp_model_temp.error<icp_model_final.error){
            icp_model_final = icp_model_temp;
        }        
    }

    // if skipping articulations on first, refit best fit    
    if(artic_svd_inital && model.joints.size()>0){
        icp_model_final = icp(model,target,treePtr,n,false, icp_model_final.R, icp_model_final.t, icp_model_final.angles, icp_model_final.scale, diameter);
    }
    return icp_model_final;
}

Eigen::Matrix3Xd get_cloud_interest(Eigen::Matrix3Xd scene, Eigen::Vector3d ref_point, double diameter){
    int const kMaxLeafCount = 16;
    std::vector<pico_tree::Neighbor<int, Eigen::Matrix3Xd::Scalar>> nn;
    pico_tree::KdTree<pico_tree::EigenTraits<Eigen::Matrix3Xd>> tree(scene, kMaxLeafCount);
    tree.SearchRadius(ref_point, diameter/2.0, &nn);

    std::vector<int> indices;     
    for(size_t i = 0; i<nn.size(); i++){
    indices.push_back(nn[i].index);
    }

    return scene(Eigen::placeholders::all,indices);
}

File: Execution/affordance_corrections/nodes/affordance_corrections/affordance_helpers/registration.py
#!/usr/bin/env python3
# File name: registration.py
# Description: register a model using ICP
# Author: Mike Hagenow
# Date: 6/7/2021

import argparse
import numpy as np
from scipy.spatial import cKDTree as KDTree
from scipy.stats import special_ortho_group
from affordance_corrections.affordance_helpers.urdf_helpers import getPointsWithAngles
from affordance_corrections.affordance_helpers.pcl_helpers import get_cloud_of_interest, get_cloud_of_interest_np
from scipy.optimize import minimize
from scipy.spatial.transform import Rotation as ScipyR
import time
import matplotlib.pyplot as plt

class FitInfo:
    ''' used to keep track of the details of a particular model's fit
        also used when pose, etc get corrections to store updates '''
    def __init__(self,pos,rot,angles,scale,residual):
        self.pos = pos
        self.rot = rot
        self.angles = angles
        self.scale = scale
        self.residual = residual

def get_fits_for_models(scene,ref_point,models,artic_svd_initial):
    """ Runs ICP random restarts for each of the provided models and returns a list of the fits for each model """
    fits = []
    for ii in range(0,len(models)):
        cloud_trimmed = get_cloud_of_interest(scene,ref_point,models[ii].diameter)
        R_final,t_final,angles_final,scale_final,min_error = icp_random_restarts(models[ii],cloud_trimmed,25,100,diameter=models[ii].diameter,ref_pt = ref_point, artic_svd_initial=artic_svd_initial)
        info_temp = FitInfo(t_final,R_final,angles_final,scale_final,min_error)
        fits.append(info_temp)
    return fits

def icp_random_restarts(model,target,num_restarts,n, diameter,ref_pt, artic_svd_initial):
    """Run ICP multiple times to avoid local minima (random rotations, translations, angles, scale)."""
    min_error = np.inf
    # Store target in KDTree for fast lookup of closest points
    tree_target = KDTree(target)
    
    # Default the return values in case nothing is found (e.g., clicked off point cloud)
    R_final = np.eye(3)
    t_final = ref_pt
    angles_final = [0 for joint in model.joints]

    for ii in range(0,num_restarts):
        # Get random start for orientation, translation, and all relevant ariculations
        R_init = special_ortho_group.rvs(3)
        # t_init = ((diameter/2.0)*np.random.random_sample()*special_ortho_group.rvs(3)[0,:]+ref_pt).reshape((3,))
        t_init = ((diameter/2.0)*np.random.normal(0.0, 0.33, 1)*special_ortho_group.rvs(3)[0,:]+ref_pt).reshape((3,))
        
        angles = []
        for ii in range(0,len(model.joints)): # model articulations
            angles.append(model.joints[ii].min_angle + (model.joints[ii].max_angle-model.joints[ii].min_angle)*np.random.rand())

        scale_init = 1.0
        if model.scaling_allowed:
            scale_init = model.min_scale + (model.max_scale-model.min_scale)*np.random.rand()

        R_temp,t_temp,angles_temp,scale_temp,error_temp = icp(model, target, tree_target, n, R_initial=R_init,t_initial=t_init, angles_initial = angles, scale_initial = scale_init, diameter = diameter, artic_svd_initial=artic_svd_initial)
        
        # Choose whether to normalize mesh size based on the approx size of the particular model
        error_temp = error_temp
        # error_temp = error_temp/model.volume # normalize based on mesh size

        if error_temp<min_error:
            R_final = R_temp
            t_final = t_temp
            angles_final = angles_temp
            scale_final = scale_temp
            min_error = error_temp

    # if skipping articulations on first, refit best fit
    if artic_svd_initial and len(model.joints)>0:
        R_final,t_final,angles_final,scale_final,min_error = icp(model, target, tree_target, n, R_initial=R_final,t_initial=t_final, angles_initial = angles_final, scale_initial = scale_final, diameter = diameter, artic_svd_initial=False)

    return R_final,t_final,angles_final,scale_final,min_error

def get_closest_alignment(aligned_source,target,target_kdtree,max_distance_threshold):
    ''' Function for aligning closest points between source and target to be used for SVD'''
    
    if len(target)==0: # no points
        return np.zeros((0,3)), np.zeros((0,3)), np.zeros((0,3)), []
    dists, indices = target_kdtree.query(aligned_source,k=1)
    closest_target = target[indices]

    # Use mask filter to get the final sets of points that should be compared
    close_enough_inds = dists < max_distance_threshold # remove outliers
    s_vals = aligned_source[close_enough_inds]
    t_vals = closest_target[close_enough_inds]

    return s_vals, t_vals, dists, close_enough_inds

def registration_residual(x,model,t_vals,close_enough_inds, weights, R_final, trans_final):
    ''' Calculates a residual for nonlinear fitting of R, t, angles, and scale if applicable '''
    
    # optimization variables are: rotation (4- quat), translation (3), angles (if app.), scale (if app.)
    R = ScipyR.from_quat(x[0:4]).as_matrix()
    t = x[4:7]
    angles_final = list(x[7:7+len(model.joints)])

    if model.scaling_allowed:
        scale = x[-1]
    else:
        scale = 1.0

    R_temp = R @ R_final
    trans_temp = R @ trans_final + t

    aligned_source = getPointsWithAngles(model,angles_final,R_temp,trans_temp, scale)

    s_vals = aligned_source[close_enough_inds] # using previous alignment (point indices)

    non_weighted_residual = np.linalg.norm(s_vals - t_vals,axis=1).flatten()
    return np.dot(weights,non_weighted_residual)/scale # normalize by scale of model to avoid scale heading towards zero

def quat_con(x):
    ''' quaternion must have magnitude 1 -- used in NL optimization '''
    return 1.0-np.linalg.norm(x[0:4])

def icp(model,target,target_kdtree,n, R_initial=np.eye(3),t_initial = np.zeros((3,)), angles_initial=[], scale_initial =1.0 , diameter = 1.0, artic_svd_initial = False, refitting = False):
    """Take two arrays of points in 3d space and computes the registration."""
    """ source refers to mesh/model to be fit, target refers to scene/point cloud"""
    """ source and target dimensions are ax3 and bx3 """
    """ Note: target should be a region of interest (~ centered wrt the selected point) """

    converged = False
    max_distance_threshold = diameter
    num_iters = 0

    R_final = R_initial
    trans_final = t_initial
    angles_final = angles_initial
    prev_error = np.inf
    scale_final = scale_initial

    # Get initial sampling with angles

    aligned_source = getPointsWithAngles(model,angles_final,R_final,trans_final,scale_final)

    

    while not converged and num_iters<n:
        if(num_iters==0):
            

            s_vals, t_vals, dists, close_enough_inds = get_closest_alignment(aligned_source,target,target_kdtree,max_distance_threshold)
            if refitting:
                print("PYTHON REFITTING")
                plt.hist(dists, density=True, cumulative=True, label='CDF', histtype='step', alpha=0.8, color='k')
                plt.show()
                print("New distance threshold: ")
                max_distance_threshold = float(input())
                s_vals, t_vals, dists, close_enough_inds = get_closest_alignment(aligned_source,target,target_kdtree,max_distance_threshold)
                aligned_source = s_vals
                max_distance_threshold = 999.9

            error = np.linalg.norm((R_initial @ s_vals.T).T+t_initial - t_vals)
            # print("Initial error: ",error)
            if len(s_vals)<3: # need 3 points
                # insufficient points for registration, return initial pose
                return R_final,trans_final,angles_final, scale_final, np.inf # return as failure (infinite error)

        

        # Weighting is based on the inverse of distances (with some normalization)
        weights = 1/(1+np.power(dists[close_enough_inds],10.0))
        weights_matrix = np.diag(weights)

        if (not model.requires_NL_fitting() or artic_svd_initial): # use SVD if not fitting articulations/scale
            # Following this formulation: https://igl.ethz.ch/projects/ARAP/svd_rot.pdf
            w_cent_s = np.divide(weights @ s_vals ,np.sum(weights))
            w_cent_t = np.divide(weights @ t_vals ,np.sum(weights))

            s_vals -= w_cent_s
            t_vals -= w_cent_t

            S = s_vals.T @ weights_matrix @ t_vals
            U, sigma, Vh = np.linalg.svd(S)
            V = np.transpose(Vh)
            R = V @ U.T
            trans = (w_cent_t - R @ w_cent_s)
            
            # print("WC: ",w_cent_t,w_cent_s)
            # print("trs: ",trans)

            R_final = R @ R_final
            trans_final = R @ trans_final + trans
        
        else: # articulations, need to use NL optimization
            cons = {'type':'eq', 'fun': quat_con}
            
            weights_new = weights

            # structure of minimization: R as quaternion, t as (3x), angles as list
            # initialize parameters
            x0 = np.zeros((7+len(angles_final)+int(model.scaling_allowed)))
            quat_temp = np.array([0.0, 0.0, 0.0, 1.0])
            x0[0:4]= quat_temp
            x0[4:7] = np.zeros((3,))
            x0[7:(7+len(angles_final))] = np.array(angles_final)
            
            if model.scaling_allowed:
                x0[-1] = scale_final

            # bound for pose (quaterions and translation)
            bounds = [(-1.0, 1.0),(-1.0, 1.0),(-1.0, 1.0),(-1.0, 1.0),(-model.diameter/2,model.diameter/2),(-model.diameter/2,model.diameter/2),(-model.diameter/2,model.diameter/2)]
    
            # Add bounds for all articulations
            for ii in range(0,len(angles_final)):
                bounds.append((model.joints[ii].min_angle,model.joints[ii].max_angle))

            # Add bounds for scaling if enabled
            if model.scaling_allowed:
                bounds.append((model.min_scale,model.max_scale))
            res = minimize(registration_residual, x0, constraints=cons, bounds=bounds,args=(model,t_vals,close_enough_inds, weights_new, R_final,trans_final), method='SLSQP', options={'disp': False,'maxiter': 50})

            #########
            # Capture output from minimization
            #########
            R = ScipyR.from_quat(res.x[0:4] ).as_matrix()
            trans = res.x[4:7]
            angles_final = list(res.x[7:7+len(angles_final)])
            # print("x0: ",x0," r:",res.x)

            if model.scaling_allowed:
                scale_final = res.x[-1]
            
            R_final = R @ R_final
            trans_final = R @ trans_final + trans
            
        # Transform source points
        aligned_source = getPointsWithAngles(model,angles_final,R_final,trans_final,scale_final)

        #########################################################################
        # Check if done - calculate residual based on next set of pt alignments #
        #########################################################################

        s_vals, t_vals, dists, close_enough_inds = get_closest_alignment(aligned_source,target,target_kdtree,max_distance_threshold)
        if len(s_vals)<3: # need 3 points
            print("Exiting early")
            # insufficient points for registration, return initial pose
            return R_final,trans_final,angles_final, np.inf # return as failure (infinite error)

        # Calculate residual
        error = np.linalg.norm((R @ s_vals.T).T+trans - t_vals)

        # Check if converged (changing by small amount -- empiricially determined)
        error_per_pt = 0.000000025
        if np.abs(prev_error-error)<(error_per_pt*len(s_vals)):
            converged = True
        prev_error = error
        num_iters+=1

        # print("end error: ",error)

    return R_final,trans_final, angles_final, scale_final, error
